settings: Add organization setting to show message moves but not message edits.

saubhagya-patel opened the pull request:
<!-- Describe your pull request here.-->

This PR adds an organization setting to customize the visible message edit history for users. 
As mentioned in the [issue](https://github.com/zulip/zulip/issues/21398):
> At present, we have a setting for whether an organization chooses to "Enable message edit history". However, some organizations may want to enable the history of messages being moved, without showing the history of messages being edited.

Fixes: #21398.
[CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/redirects.20from.20near.20links/near/1343095)

**Previous PR reference:**
- This PR continues the unfinished work done in PR [#21562](https://github.com/zulip/zulip/pull/21562) by @madrix01.
- The updates made are as follows:
   - Split the work into two commits as suggested [here](https://github.com/zulip/zulip/pull/21562#issuecomment-1098408540).
   - Made a function named `visible_edit_history_for_message()` in `zerver/lib/message.py` as suggested [here](https://github.com/zulip/zulip/pull/21562#discussion_r849822349).
   - Implement tests for the above function.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Added setting for message edit history.</summary>

**Setting: Show message edit history?**
![image](https://github.com/user-attachments/assets/cda713b9-4002-4f67-92c2-2be081a17c54)

**Dropdown**
![image](https://github.com/user-attachments/assets/cb01b332-d3ce-47da-95fb-6789fed365b5)

</details>

<details>
<summary>Show message edit history - Always.</summary>

![image](https://github.com/user-attachments/assets/2359a5d6-d434-41e2-8bc6-f82e9ac37feb)

</details>

<details>
<summary>Show message edit history - Moves only.</summary>

![image](https://github.com/user-attachments/assets/9862def5-8f73-46db-8534-e85a55fb9c64)

</details>

<details>
<summary>Show message edit history - Never.</summary>

Please note that only the tooltip appears here since we don't fetch and render the message edit history.
![image](https://github.com/user-attachments/assets/2884d5b0-f2ae-4593-8836-b30ef8ead6e3)

</details>

<details>
<summary>Side-by-side screenshots for better comparison.</summary>

Please note that in "Show message edit history - Never" only the tooltip appears since we don't fetch and render the message edit history.
| Show message edit history - Always | Show message edit history - Moves only | Show message edit history - Never |
|-------------------------------------|----------------------------------------|----------------------------------|
| ![image](https://github.com/user-attachments/assets/2359a5d6-d434-41e2-8bc6-f82e9ac37feb) | ![image](https://github.com/user-attachments/assets/9862def5-8f73-46db-8534-e85a55fb9c64) | ![image](https://github.com/user-attachments/assets/2884d5b0-f2ae-4593-8836-b30ef8ead6e3) |


</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
@sahil839 are you up for reviewing this one? I haven't tested.

alya commented:
I also put up #32886 after seeing these screenshots.

sahil839 commented on zerver/openapi/zulip.yaml:
```diff
@@ -4770,6 +4765,14 @@ paths:
+                                        Which type of history i.e. moved or edited is configured to allow users to access
+                                        [message edit history](/help/view-a-messages-edit-history).
+
+                                        **Changes**: New in Zulip 10.0 (feature level 333), `allow_edit_history` setting
```
I think this text can also be phrased like below.

saubhagya-patel replied:
Sure, it is fixed in subsequent push.

sahil839 commented on zerver/tests/test_message_edit.py:
```diff
@@ -905,6 +914,83 @@ def test_edit_cases(self) -> None:
+        message_edit_history_none = visible_edit_history_for_message(
+            MessageEditHistoryVisibilityEnum.MESSAGE_EDIT_HISTORY_VISIBILITY_NONE, message_history
+        )
+        self.assert_length(message_edit_history_none, 0)
```
Shouldn't this just check the `message_history` field in response?

saubhagya-patel replied:
This test was written to verify the visible edit history returned by the `visible_edit_history_for_message()` function.
The functions `test_edit_message_history_disabled()` and `test_edit_message_history()` check the `message_history` field in response. So, the `message_history` field in response was not checked here. I hope its appropriate.

sahil839 replied:
It would be better to just test the `message_history` field in response for all the possible setting values instead of testing this function, since this function is eventually used to get the value for `message_history` field.

saubhagya-patel commented on zerver/migrations/0644_realm_message_edit_history_visibility.py:
```diff
@@ -0,0 +1,37 @@
+        migrations.RunPython(migrate_allow_edit_history),
+        migrations.RemoveField(
+            model_name="realm",
+            name="allow_edit_history",
```
Oh, I didn't know about that. I will fix this and thanks for the info.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:
- Changed the `zulip.yaml` file following [this](https://github.com/zulip/zulip/pull/32840#discussion_r1903240757) comment.
- Separated migration files as stated in [this](https://github.com/zulip/zulip/pull/32840#discussion_r1903215136) comment.
- No visual changes are made.
- On testing locally, I found a bug where if we send a message to a topic -> move it to a different topic -> edit the message -> view its edit history (when moves only message edit history is selected) it raises an Assertion error, now, it is fixed.

@sahil839, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further adjustments are needed.
Thank you.

sahil839 commented on web/src/message_events.ts:
```diff
@@ -417,7 +418,10 @@ export function update_messages(events: UpdateMessageEvent[]): void {
+                if (
+                    realm.realm_message_edit_history_visibility !==
+                    message_edit_history_visibility_values.never.code
+                ) {
```
Should this condition whether edit history visibilty is set to "All"? I guess there is no need for the below code to execute if setting is set to only show moves history.

saubhagya-patel replied:
That sounds reasonable. Fixed it. Thank you.

sahil839 commented on zerver/migrations/0646_set_value_for_message_edit_history_visibility.py:
```diff
@@ -0,0 +1,28 @@
+    )
+
+
+class Migration(migrations.Migration):
```
This might need `atomic = False` as `Realm` table can be large for Zulip Cloud.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes are made:

- Updated the `test_visible_edit_history_for_message()` function in `test_message_edit.py` file to check the `message_history` field in response for all the setting values as stated in [this](https://github.com/zulip/zulip/pull/32840#discussion_r1906589968) comment.

- Updated code in `message_events.ts` as stated in [this](https://github.com/zulip/zulip/pull/32840#discussion_r1906593151) comment.

- Added `atomic = False` to `0650_set_value_for_message_edit_history_visibility.py` migration file as stated in [this](https://github.com/zulip/zulip/pull/32840#discussion_r1906603666) comment.

- In the `visible_edit_history_for_message()` function in `zerver/lib/message.py` the check for message edit history->none was removed as it was redundant.

- No visual changes are made.

@sahil839, this PR is ready for review. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

sahil839 commented on zerver/tests/test_message_edit.py:
```diff
@@ -907,6 +916,140 @@ def test_edit_cases(self) -> None:
+        for edit_history_entry in message_edit_history_moves_only:
+            self.assertNotIn("prev_content", edit_history_entry)
+            self.assertNotIn("prev_rendered_content", edit_history_entry)
+            self.assertNotIn("content_html_diff", edit_history_entry)
```
I think this should also check the entries present in the history dict individually.

sahil839 replied:
And probably a check for length of history entries list as well would be good.

saubhagya-patel replied:
These changes are included in the subsequent push

sahil839 commented on zerver/tests/test_message_edit.py:
```diff
@@ -907,6 +916,140 @@ def test_edit_cases(self) -> None:
+            if i in {3, 5}:
+                expected_entries.add("prev_stream")
+                expected_entries.add("stream")
+            self.assertEqual(expected_entries, set(entry.keys()))
```
I am not sure having these checks is helpful. Would be just better to have a check for length and to verify the values in those entries.

sahil839 reviewed (commented):
Posted a couple of comments. Looks good otherwise. Can test it once you rebase the PR to resolve merge conflicts.

sahil839 commented:
I tested and it looks good.

saubhagya-patel commented:
@timabbott, a gentle ping for this PR. Thank you.

timabbott commented:
Keep this updated, I'll try to do a proper integration review tomorrow.

timabbott commented on zerver/lib/message.py:
```diff
@@ -210,14 +211,35 @@ def truncate_topic(topic_name: str) -> str:
+    for edit_history_event in edit_history:
+        if "prev_content" in edit_history_event:
+            continue
+        visible_edit_history.append(edit_history_event)
```
Hmm, what if an edit history event contains both `prev_content` and `prev_topic`? I'm sure that's very rare, but I think it used to be possible. It would likely be correct to drop the `content/prev_content` fields from such events, rather than skipping them entirely, since the message move part should still be accessible to clients, and in fact needs to be there for the chain of `prev_topic` values to be correct.

You can likely implement some unit tests for that sort of conversion.

BTW, as a side note, https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history/near/2057755 would be a great project for you to pick up next.

saubhagya-patel commented on zerver/openapi/zulip.yaml:
```diff
@@ -4576,11 +4576,6 @@ paths:
-                                      type: boolean
-                                      description: |
-                                        Whether this organization is configured to allow users to access
-                                        [message edit history](/help/view-a-messages-edit-history).
```
Updated the `POST /register` response for backwards-compatibility following the [CZO thread](https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history.20visibility/near/2091730).

saubhagya-patel commented:
I have updated the PR, following the review received. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32840#discussion_r1956666641) comment regarding if an edit history event contains both `prev_content` and `prev_topic`, the `visible_edit_history_for_message()` function in `zerver/lib/message.py` and corresponding tests in the `test_visible_edit_history_for_message()` function in `zerver/tests/test_message_edit.py` are updated.

- Following [this](https://github.com/zulip/zulip/pull/32840#discussion_r1956676586) comment and the [CZO thread](https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history.20visibility/near/2091730) regarding the `POST /register` response for backwards-compatibility, the endpoint is updated. The `POST /register` response now includes `realm_allow_edit_history` field as it is used by the production mobile help as suggested over [here](https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history.20visibility/near/2091989).
  
  - The value of `realm_allow_edit_history` field is set to `False` if the `realm_message_edit_history_visibility_policy` is "None", and `True` for "Moves only" or "All" message edit history as discussed in the [CZO thread](https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history.20visibility/near/2091730).


Please let me know if I missed anything or if any adjustments are needed.
Thank you.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -16852,13 +16860,36 @@ paths:
+                          **Changes**: Deprecated in Zulip 10.0 (feature level 354) and replaced
+                          by `realm_message_edit_history_visibility_policy` which supports an
+                          intermediate field of `Moves only` history.
+                          We set the value of `realm_allow_edit_history` as `false` if the
+                          `realm_message_edit_history_visibility_policy` is configured as "None"
+                          and `true` if it is configured as "Moves only" or "All".
```
This information should go above the "Changes" note, because it's information that's needed for understanding the behavior of new servers.

In general, a reader who's only interested in understanding the latest version of the server should always be able to stop reading when they reach "**Changes**:".

saubhagya-patel replied:
The information has been moved above the "Changes" note, in the subsequent push.
Thanks for the info, will remember this in the future.

alya commented:
Let's tweak the strings in the UI; I think we can make it more self-explanatory regardless of the option that's selected.

Allow viewing the history of a message?
- Show edits and moves
- Move history only
- Don't allow

timabbott commented on zerver/actions/message_summary.py:
```diff
@@ -120,7 +121,7 @@ def do_summarize_narrow(
         allow_empty_topic_name=False,
         # Avoid fetching edit history, which won't be passed to the model.
-        allow_edit_history=False,
+        message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.MESSAGE_EDIT_HISTORY_VISIBILITY_NONE,
```
This is a bit absurdly long. I think the thing we can do is this -- no need for the values within the `Enum`, which are already namespaced, to repeat the type:

`message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE`

saubhagya-patel commented:
Following the review received, I have updated the PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/32840#issuecomment-2666287519) comment, the strings in the UI have been updated. Kindly reference the attached screenshots for the visual changes that were made.

- Following [this](https://github.com/zulip/zulip/pull/32840#discussion_r1961035816) comment, the values in the `MessageEditHistoryVisibilityPolicyEnum` are renamed.

- Visual changes: 

| Before | After |
|--------------|-------------|
| ![test](https://github.com/user-attachments/assets/a129eb69-c798-490d-b6a3-df78700b802e) | ![image](https://github.com/user-attachments/assets/94c7af43-79d9-4e7b-a399-64bcbc9c8cb1) |

@alya, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

timabbott commented:
I figured out how to make this use the enum format proposed:

``` diff
diff --git a/zerver/actions/message_summary.py b/zerver/actions/message_summary.py
index 2f612d5985..b1669dd744 100644
--- a/zerver/actions/message_summary.py
+++ b/zerver/actions/message_summary.py
@@ -121,7 +121,7 @@ def do_summarize_narrow(
         client_gravatar=True,
         allow_empty_topic_name=False,
         # Avoid fetching edit history, which won't be passed to the model.
-        message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+        message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
         user_profile=user_profile,
         realm=user_profile.realm,
     )
diff --git a/zerver/actions/realm_settings.py b/zerver/actions/realm_settings.py
index ae5d276f67..d14e5a26b0 100644
--- a/zerver/actions/realm_settings.py
+++ b/zerver/actions/realm_settings.py
@@ -2,6 +2,7 @@ import datetime
 import logging
 import zoneinfo
 from email.headerregistry import Address
+from enum import Enum
 from typing import Any, Literal
 
 from django.conf import settings
@@ -57,17 +58,22 @@ from zerver.tornado.django_api import send_event_on_commit
 
 @transaction.atomic(savepoint=False)
 def do_set_realm_property(
-    realm: Realm, name: str, value: Any, *, acting_user: UserProfile | None
+    realm: Realm, name: str, raw_value: Any, *, acting_user: UserProfile | None
 ) -> None:
     """Takes in a realm object, the name of an attribute to update, the
     value to update and the user who initiated the update.
     """
     property_type = Realm.property_types[name]
-    assert isinstance(value, property_type), (
-        f"Cannot update {name}: {value} is not an instance of {property_type}"
+    assert isinstance(raw_value, property_type), (
+        f"Cannot update {name}: {raw_value} is not an instance of {property_type}"
     )
 
     old_value = getattr(realm, name)
+    if isinstance(raw_value, Enum):
+        value = raw_value.value
+    else:
+        value = raw_value
+
     if old_value == value:
         return
 
diff --git a/zerver/lib/event_schema.py b/zerver/lib/event_schema.py
index bb71d4b4ad..3efbdbe1b6 100644
--- a/zerver/lib/event_schema.py
+++ b/zerver/lib/event_schema.py
@@ -6,6 +6,8 @@
 #
 # See https://zulip.readthedocs.io/en/latest/subsystems/events-system.html
 from collections.abc import Callable
+from enum import Enum
+import inspect
 from pprint import PrettyPrinter
 from typing import cast
 
@@ -426,7 +428,10 @@ def check_realm_update(
         return
 
     property_type = Realm.property_types[prop]
-    assert isinstance(value, property_type)
+    if inspect.isclass(property_type) and issubclass(property_type, Enum):
+        assert isinstance(value, int)
+    else:
+        assert isinstance(value, property_type)
 
 
 def check_realm_default_update(
diff --git a/zerver/lib/events.py b/zerver/lib/events.py
index b79971983a..dbd2525909 100644
--- a/zerver/lib/events.py
+++ b/zerver/lib/events.py
@@ -493,7 +493,7 @@ def fetch_initial_state_data(
 
         state["realm_allow_edit_history"] = (
             realm.message_edit_history_visibility_policy
-            != MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE
+            != MessageEditHistoryVisibilityPolicyEnum.none.value
         )
 
     if want("realm_user_settings_defaults"):
@@ -1336,7 +1336,7 @@ def apply_event(
             if field == "realm_message_edit_history_visibility_policy":
                 state["realm_allow_edit_history"] = (
                     user_profile.realm.message_edit_history_visibility_policy
-                    != MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE
+                    != MessageEditHistoryVisibilityPolicyEnum.none.value
                 )
 
         elif event["op"] == "update_dict":
diff --git a/zerver/lib/message.py b/zerver/lib/message.py
index f93c8dbfdc..02602abf36 100644
--- a/zerver/lib/message.py
+++ b/zerver/lib/message.py
@@ -227,7 +227,7 @@ def visible_edit_history_for_message(
     # in realms as per `message_edit_history_visibility_policy`.
     if (
         message_edit_history_visibility_policy
-        == MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_ALL
+        == MessageEditHistoryVisibilityPolicyEnum.all.value
     ):
         return edit_history
 
@@ -289,7 +289,7 @@ def messages_for_ids(
         if "edit_history" in msg_dict:
             if (
                 message_edit_history_visibility_policy
-                == MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE
+                == MessageEditHistoryVisibilityPolicyEnum.none.value
             ):
                 del msg_dict["edit_history"]
             else:
diff --git a/zerver/models/realms.py b/zerver/models/realms.py
index 2b5f1db1a0..819f651abf 100644
--- a/zerver/models/realms.py
+++ b/zerver/models/realms.py
@@ -1,5 +1,5 @@
 from email.headerregistry import Address
-from enum import IntEnum
+from enum import Enum, IntEnum
 from types import UnionType
 from typing import TYPE_CHECKING, Optional, TypedDict
 from uuid import uuid4
@@ -148,10 +148,11 @@ class DigestWeekdayEnum(IntEnum):
     SUNDAY = 6
 
 
-class MessageEditHistoryVisibilityPolicyEnum(IntEnum):
-    VISIBILITY_ALL = 1
-    VISIBILITY_MOVES_ONLY = 2
-    VISIBILITY_NONE = 3
+class MessageEditHistoryVisibilityPolicyEnum(Enum):
+    # The case is used by Pydantic in the API
+    all = 1
+    moves = 2
+    none = 3
 
 
 class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stubs cannot resolve the custom CTEManager yet https://github.com/typeddjango/django-stubs/issues/1023
@@ -397,10 +398,10 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
 
     # Whether users have access to message edit history
     message_edit_history_visibility_policy = models.PositiveSmallIntegerField(
-        default=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_ALL,
+        default=MessageEditHistoryVisibilityPolicyEnum.all.value,
     )
     MESSAGE_EDIT_HISTORY_VISIBILITY_POLICY_TYPES = [
-        field.value for field in MessageEditHistoryVisibilityPolicyEnum
+        field for field in MessageEditHistoryVisibilityPolicyEnum
     ]
 
     # Defaults for new users
@@ -685,7 +686,7 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
         message_content_allowed_in_email_notifications=bool,
         message_content_edit_limit_seconds=int | None,
         message_content_delete_limit_seconds=int | None,
-        message_edit_history_visibility_policy=int,
+        message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum,
         move_messages_between_streams_limit_seconds=int | None,
         move_messages_within_stream_limit_seconds=int | None,
         message_retention_days=int,
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 963be28da6..333a6ecba9 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -5056,18 +5056,18 @@ paths:
                                         **Changes**: Before Zulip 6.0 (feature level 138), no limit was
                                         represented using the special value `0`.
                                     message_edit_history_visibility_policy:
-                                      type: integer
+                                      type: string
                                       description: |
                                         Which type of message edit history is configured to allow users to
                                         access [message edit history](/help/view-a-messages-edit-history).
 
-                                        - 1 = All history is visible
-                                        - 2 = Only Moves history is visible
-                                        - 3 = No history is visible
+                                        - "all" = All edit history is visible.
+                                        - "moves" = Only moves are visible.
+                                        - "none" = No edit history is visible.
 
                                         **Changes**: New in Zulip 10.0 (feature level 355), replacing the previous
-                                        `allow_edit_history` boolean setting; `true` corresponds to `All`,
-                                        and `false` to `None`.
+                                        `allow_edit_history` boolean setting; `true` corresponds to `all`,
+                                        and `false` to `none`.
                                     moderation_request_channel_id:
                                       type: integer
                                       description: |
@@ -16863,18 +16863,16 @@ paths:
                       realm_message_edit_history_visibility_policy:
                         type: integer
                         description: |
-                          Present if `realm` is present in `fetch_event_types`.
-
-                          Which type of message edit history is configured to allow users to
-                          access [message edit history](/help/view-a-messages-edit-history).
+                          What typesof message edit history are accessible to users via
+                          [message edit history](/help/view-a-messages-edit-history).
 
-                          - 1 = All history is visible
-                          - 2 = Only Moves history is visible
-                          - 3 = No history is visible
+                          - "all" = All edit history is visible.
+                          - "moves" = Only moves are visible.
+                          - "none" = No edit history is visible.
 
-                          **Changes**: New in Zulip 10.0 (feature level 355) replacing the previous
-                          `realm_allow_edit_history` boolean setting; `true` corresponds to `All`,
-                          and `false` to `None`.
+                          **Changes**: New in Zulip 10.0 (feature level 355), replacing the previous
+                          `allow_edit_history` boolean setting; `true` corresponds to `all`,
+                          and `false` to `none`.
                       realm_allow_edit_history:
                         deprecated: true
                         type: boolean
diff --git a/zerver/tests/test_events.py b/zerver/tests/test_events.py
index 0bf4f9ad1f..f722b55ea0 100644
--- a/zerver/tests/test_events.py
+++ b/zerver/tests/test_events.py
@@ -9,6 +9,7 @@ import time
 from collections.abc import Iterator
 from contextlib import contextmanager
 from datetime import timedelta
+from enum import Enum
 from io import StringIO
 from typing import Any
 from unittest import mock
@@ -3876,7 +3877,9 @@ class RealmPropertyActionTest(BaseAction):
 
         do_set_realm_property(self.user_profile.realm, name, vals[0], acting_user=self.user_profile)
 
-        if vals[0] != original_val:
+        if vals[0] != original_val and not (
+            isinstance(vals[0], Enum) and vals[0].value == original_val
+        ):
             self.assertEqual(
                 RealmAuditLog.objects.filter(
                     realm=self.user_profile.realm,
@@ -3886,11 +3889,18 @@ class RealmPropertyActionTest(BaseAction):
                 ).count(),
                 1,
             )
-        for count, val in enumerate(vals[1:]):
+        for count, raw_value in enumerate(vals[1:]):
             now = timezone_now()
             state_change_expected = True
-            old_value = vals[count]
             num_events = 1
+            raw_old_value = vals[count]
+
+            if isinstance(raw_value, Enum):
+                value = raw_value.value
+                old_value = raw_old_value.value
+            else:
+                value = raw_value
+                old_value = raw_old_value
 
             with self.verify_action(
                 state_change_expected=state_change_expected, num_events=num_events
@@ -3898,7 +3908,7 @@ class RealmPropertyActionTest(BaseAction):
                 do_set_realm_property(
                     self.user_profile.realm,
                     name,
-                    val,
+                    raw_value,
                     acting_user=self.user_profile,
                 )
 
@@ -3910,7 +3920,7 @@ class RealmPropertyActionTest(BaseAction):
                     acting_user=self.user_profile,
                     extra_data={
                         RealmAuditLog.OLD_VALUE: old_value,
-                        RealmAuditLog.NEW_VALUE: val,
+                        RealmAuditLog.NEW_VALUE: value,
                         "property": name,
                     },
                 ).count(),
diff --git a/zerver/tests/test_message_dict.py b/zerver/tests/test_message_dict.py
index 539ea79704..9482d4a706 100644
--- a/zerver/tests/test_message_dict.py
+++ b/zerver/tests/test_message_dict.py
@@ -421,7 +421,7 @@ class MessageHydrationTest(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=cordelia,
             realm=cordelia.realm,
         )
@@ -470,7 +470,7 @@ class MessageHydrationTest(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=self.example_user("polonius"),
             realm=realm,
         )
@@ -516,7 +516,7 @@ class MessageHydrationTest(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=cordelia,
             realm=cordelia.realm,
         )
@@ -562,7 +562,7 @@ class TestMessageForIdsDisplayRecipientFetching(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=cordelia,
             realm=cordelia.realm,
         )
@@ -586,7 +586,7 @@ class TestMessageForIdsDisplayRecipientFetching(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=cordelia,
             realm=cordelia.realm,
         )
@@ -611,7 +611,7 @@ class TestMessageForIdsDisplayRecipientFetching(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=cordelia,
             realm=cordelia.realm,
         )
@@ -648,7 +648,7 @@ class TestMessageForIdsDisplayRecipientFetching(ZulipTestCase):
             apply_markdown=True,
             client_gravatar=True,
             allow_empty_topic_name=True,
-            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.none.value,
             user_profile=cordelia,
             realm=cordelia.realm,
         )
diff --git a/zerver/tests/test_message_edit.py b/zerver/tests/test_message_edit.py
index 627f480f21..fc6f7e87df 100644
--- a/zerver/tests/test_message_edit.py
+++ b/zerver/tests/test_message_edit.py
@@ -50,7 +50,7 @@ class EditMessageTest(ZulipTestCase):
                 apply_markdown=False,
                 client_gravatar=False,
                 allow_empty_topic_name=True,
-                message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_ALL,
+                message_edit_history_visibility_policy=MessageEditHistoryVisibilityPolicyEnum.all.value,
                 user_profile=None,
                 realm=msg.realm,
             )
@@ -516,7 +516,7 @@ class EditMessageTest(ZulipTestCase):
         do_set_realm_property(
             user_profile.realm,
             "message_edit_history_visibility_policy",
-            MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            MessageEditHistoryVisibilityPolicyEnum.none,
             acting_user=None,
         )
         self.login("hamlet")
@@ -1010,9 +1010,10 @@ class EditMessageTest(ZulipTestCase):
         do_set_realm_property(
             user.realm,
             "message_edit_history_visibility_policy",
-            MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE,
+            MessageEditHistoryVisibilityPolicyEnum.none,
             acting_user=None,
         )
+        print(user.realm.message_edit_history_visibility_policy)
         result_message_edit_history_none = self.client_get(f"/json/messages/{msg_id}/history")
         self.assert_json_error(
             result_message_edit_history_none,
@@ -1022,7 +1023,7 @@ class EditMessageTest(ZulipTestCase):
         do_set_realm_property(
             user.realm,
             "message_edit_history_visibility_policy",
-            MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_MOVES_ONLY,
+            MessageEditHistoryVisibilityPolicyEnum.moves,
             acting_user=None,
         )
         result_message_edit_history_moves_only = self.client_get(f"/json/messages/{msg_id}/history")
@@ -1060,7 +1061,7 @@ class EditMessageTest(ZulipTestCase):
         do_set_realm_property(
             user.realm,
             "message_edit_history_visibility_policy",
-            MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_ALL,
+            MessageEditHistoryVisibilityPolicyEnum.all,
             acting_user=None,
         )
         result_message_edit_history_all = self.client_get(f"/json/messages/{msg_id}/history")
diff --git a/zerver/views/message_edit.py b/zerver/views/message_edit.py
index e55596d463..ca8270e49d 100644
--- a/zerver/views/message_edit.py
+++ b/zerver/views/message_edit.py
@@ -116,7 +116,7 @@ def get_message_edit_history(
     )
     if (
         user_realm_message_edit_history_visibility_policy
-        == MessageEditHistoryVisibilityPolicyEnum.VISIBILITY_NONE
+        == MessageEditHistoryVisibilityPolicyEnum.none.value
     ):
         raise JsonableError(_("Message edit history is disabled in this organization"))
     message = access_message(user_profile, message_id)
```

timabbott commented:
I pushed those changes back here -- see https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history.20visibility/near/2100698 for discussion on whether we have the right API model for this, though.

timabbott commented:
I think this is just waiting on fixing the outgoing register/events format to use the strings rather than integers in the API.

saubhagya-patel commented:
Following the changes pushed, I have updated the PR.
Here, is a summary of the changes made: 

- The API feature levels are presented following the [#api design > API feature levels rebasing idea](https://chat.zulip.org/#narrow/channel/378-api-design/topic/API.20feature.20levels.20rebasing.20idea/with/2101911) as stated [here]([#api design > edit history visibility @ 💬](https://chat.zulip.org/#narrow/channel/378-api-design/topic/edit.20history.20visibility/near/2100960)). 

- Cleaned up some changes and addressed [this](https://github.com/zulip/zulip/pull/32840#issuecomment-2680798271) comment to use strings in the API.

@timabbott, this PR is ready for review.
I hope the updated changes look good to you. Please let me know if any further changes are required.
Thank you for the changes you made to use the enum format. It was really helpful.

timabbott commented:
I renumbered migrations and ran `tools/merge-api-changelogs`, which worked! @Vector73 FYI :). Merged with the following small tweaks:

``` diff
diff --git a/zerver/lib/events.py b/zerver/lib/events.py
index 47e7d6b642..4abec539de 100644
--- a/zerver/lib/events.py
+++ b/zerver/lib/events.py
@@ -1326,8 +1326,7 @@ def apply_event(
 
             if field == "realm_message_edit_history_visibility_policy":
                 state["realm_allow_edit_history"] = (
-                    user_profile.realm.message_edit_history_visibility_policy
-                    != MessageEditHistoryVisibilityPolicyEnum.none.value
+                    event["value"] != MessageEditHistoryVisibilityPolicyEnum.none.name
                 )
 
         elif event["op"] == "update_dict":
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 1557bf75d6..618ad6557a 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -16929,9 +16929,9 @@ paths:
                           `realm_message_edit_history_visibility_policy` is configured as "None"
                           and `true` if it is configured as "Moves only" or "All".
 
-                          **Changes**: Deprecated in Zulip 10.0 (feature level 358) and replaced
-                          by `realm_message_edit_history_visibility_policy` which supports an
-                          intermediate field of `Moves only` history.
+                          **Changes**: Deprecated in Zulip 10.0 (feature level 358) and will be
+                          removed in the future, as it is an inaccurate version
+                          `realm_message_edit_history_visibility_policy`, which replaces this field.
                       realm_can_add_custom_emoji_group:
                         allOf:
                           - description: |
```

I also fixed one bug, which is that there was no validation of values being valid, with a TODO that you should pick up next:
``` diff
diff --git a/zerver/tests/test_realm.py b/zerver/tests/test_realm.py
index 6c22bbd0f8..55d38f948b 100644
--- a/zerver/tests/test_realm.py
+++ b/zerver/tests/test_realm.py
@@ -2330,6 +2330,13 @@ class RealmAPITest(ZulipTestCase):
         result = self.client_patch("/json/realm", {"org_type": invalid_org_type})
         self.assert_json_error(result, "Invalid org_type")
 
+    def test_invalid_edit_history_visibility(self) -> None:
+        result = self.client_patch(
+            "/json/realm",
+            {"message_edit_history_visibility_policy": "invalid"},
+        )
+        self.assert_json_error(result, "Invalid message_edit_history_visibility_policy")
+
     def update_with_realm_default_api(self, name: str, val: Any) -> None:
         if not isinstance(val, str):
             val = orjson.dumps(val).decode()
diff --git a/zerver/views/realm.py b/zerver/views/realm.py
index b8a49a7240..ba6ecf5a11 100644
--- a/zerver/views/realm.py
+++ b/zerver/views/realm.py
@@ -118,7 +118,8 @@ def update_realm(
         Json[int | str] | None, ApiParamConfig("message_content_edit_limit_seconds")
     ] = None,
     message_edit_history_visibility_policy_raw: Annotated[
-        str | None,
+        # TODO: Use MessageEditHistoryVisibilityPolicyEnum here with Pydantic
+        Literal["all", "moves", "none"] | None,
         ApiParamConfig("message_edit_history_visibility_policy"),
     ] = None,
     default_language: str | None = None,
```

Merged, thanks for all the work on this @saubhagya-patel!
